Recognize %identity and %obj_magic as nonexpansive

stedolan opened the pull request:
This code should not hit the value restriction:
```ocaml
let f = Obj.magic f
```
This change lets the compiler see that `Obj.magic` and similar primitives aren't actually applications, so can't hide a mutable cell allocation, so don't need to be subject to the value restriction. (We already have a similar check for `raise`).

goldfirere reviewed (commented):
I suppose if I

```
let r = ref []
let r2 : 'a. 'a ref = Obj.magic r
```

I get what I deserve?

stedolan commented on testsuite/tests/typing-misc/magic_is_a_value.ml:
```diff
@@ -0,0 +1,16 @@
+(* not really an application, since Obj.magic doesn't do anything *)
+let f : 'a -> 'a = Obj.magic Fun.id
+[%%expect {|
+val f : ('a : any). 'a -> 'a = <fun>
```
Huh, the kinding of variables is not quite what I expected:
```
# let f : 'a . 'a -> 'a = Obj.magic Fun.id ;;
val f : 'a -> 'a = <fun>
# let f : 'a -> 'a = Obj.magic Fun.id;;
val f : ('a : any). 'a -> 'a = <fun>
```

goldfirere replied:
That's operating as designed. When you bind a type variable explicitly, it has kind `value` unless you request otherwise. But unification variables in a `let` are born at kind `any` and get lowered only when necessary. This avoids the need for annotations in `let`s. Except when you're doing magic, evidently (because there's nothing to lower the kind).

There's a chance, thus, that this PR would interact with existing code to create something bad. And it would allow people to do something bad more easily.

So I'm not sure I'm in favor. I'm really on the fence.

ncik-roberts replied:
I'm in favor of this PR as something that will make the prototyping of portabilizing much easier.

Notably, `Obj.magic_portable` has type `'a -> 'a @ portable`, so you don't run into the specific issue Richard is pointing out.

Maybe we could push through the change for `%identity`, even if we don't do the same for `%obj_magic`, and define `Obj.magic_portable` as `%identity`? Or maybe we could require an `[@@nonexpansive]` annotation that we add to the declaration of `Obj.magic_portable` but not other things?

ncik-roberts replied:
On second thought, I think this PR doesn't do much to change the status quo of "scariness of Obj.magic". You can already write this:

```ocaml
let f : _ -> _ = Obj.magic Obj.magic;;
```

which gives you `val f : ('a : any). '_weak1 -> 'a = <fun>`. Fine, I choose `'_weak1` is `float#`:

```
# ((f #4.0) : int);;
- : int = 69957243291688
```

goldfirere commented:
I think I'm against this as written. Even with Nick's helpful analysis above, I think this change is too likely to cause disaster, because presumably a weakly polymorphic function is actually used in the file in which it is defined (and so it won't cause disaster in practice).

I'd be fine with requiring a `[@@nonexpansive]` annotation (which we could then put on a new bit of `magic`) or excepting `%obj_magic`. And I'd be fine if we had some tooling to require annotations on uses of `Obj.magic` to have a type annotation (like we already do for e.g. `ignore`). (I would then advocate for making sure the tooling checks for wayward layouts.)

I'm also fine to be overruled here. I don't feel strongly. But after considering my stance (having been prompted offline by @stedolan), I remain against.

ncik-roberts commented:
I am not convinced by the distinction you're drawing between weak and non-weak type variables in this context. Someone still has to write `val f : ('a : any). 'a -> 'a` in an mli to get the crazy type. (And, in my example that works today, you can still expose a crazy function `val f : #(int * string) -> #(int * string)` which is plainly unsound.)

I remain of the opinion that this feature would be very useful for prototyping portability and that the concerns you raise are not so different than concerns exist today. (Maybe a greater magnitude of concern, but I'm not convinced of that either.)

stedolan commented:
After the conversation, I've updated this patch a bit:
  - rebased to current HEAD
  - per @dkalinichenko-js 's suggestion, the effect now only applies  
    `when not (Language_extension.erasable_extensions_only ())`
  - updated a couple of tests that were testing weak poly variables: `Fun.id` is no longer a reliable way to make those, so use `!(ref ...)` instead

CI is green now, so I'll merge if someone hits the "approve" button. (@goldfirere ?)

goldfirere commented on typing/typecore.ml:
```diff
@@ -4222,6 +4219,15 @@ let rec is_nonexpansive exp =
+  | ("%raise" | "%reraise" | "%raise_notrace"), [Nolabel, Arg (e, _)] ->
+    is_nonexpansive e
+  | ("%identity" | "%obj_magic"), [Nolabel, Arg (e, _)]
+    when not (Language_extension.erasable_extensions_only ()) ->
```
Maybe add a test demonstrating how the behavior of this depends on the extensions choice?

ncik-roberts replied:
I've pushed this test.

stedolan replied:
Thanks!
